R/tm_g_bivariate.R

```diff
@@ -363,41 +363,50 @@ srv_g_bivariate <- function(id,
       color = color, fill = fill, size = size
     )
 
+    rule_var <- function(other) {
+      function(value) {
+        othervalue <- selector_list()[[other]]()$select
+        if (length(value) == 0L && length(othervalue) == 0L)
+          "Please select at least one of x-variable or y-variable"
+      }
+    }
+    rule_diff <- function(other) {
+      function(value) {
+        othervalue <- selector_list()[[other]]()$select
+        if (identical(value, othervalue))
+          "Row and column facetting variables must be different."
+      }
+    }
+
     selector_list <- teal.transform::data_extract_multiple_srv(
       data_extract = data_extract,
       datasets = data,
       select_validation_rule = list(
-        x = ~ if (length(.) == 0 && length(selector_list()$y()$select) == 0)
-          "Please select at least one of x-variable or y-variable",
-        y = ~ if (length(.) == 0 && length(selector_list()$x()$select) == 0)
-          "Please select at least one of x-variable or y-variable",
+        x = rule_var("y"),
+        y = rule_var("x"),
         row_facet = shinyvalidate::compose_rules(
-          ~ if (length(.) > 1) "There must be 1 or no column facetting variable.",
-          ~ if ("col_facet" %in% names(selector_list())) {
-            if (
-              length(.) == 1 &&
-              length(selector_list()$col_facet()$select) == 1 &&
-              (.) == selector_list()$col_facet()$select)
-              "Row and column facetting variables must be different."
-          }
+          shinyvalidate::sv_optional(),
+          ~ if (length(.) > 1) "There must be 1 or no row facetting variable.",
+          crule(rule_diff("col_facet"), !is.null(input$row_facet))
         ),
         col_facet = shinyvalidate::compose_rules(
-          ~ if (length(.) > 1) "There must be 1 or no row facetting variable.",
-          ~ if ("row_facet" %in% names(selector_list())) {
-            if (
-              length(.) == 1 &&
-              length(selector_list()$row_facet()$select) == 1 &&
-              (.) == selector_list()$row_facet()$select)
-              "Row and column facetting variables must be different."
-          }
+          shinyvalidate::sv_optional(),
+          ~ if (length(.) > 1) "There must be 1 or no column facetting variable.",
+          crule(rule_diff("row_facet"), !is.null(input$col_facet))
         )
       )
     )
 
     iv_r <- reactive({
+      iv_facet <- shinyvalidate::InputValidator$new()
+      iv_child <- teal.transform::compose_and_enable_validators(iv_facet, selector_list,
+                                                                validator_names = c("row_facet", "col_facet"))
+      iv_child$condition(~ isTRUE(input$facetting))
+
```

Comment:
I wonder if this is a case for modifying `compose_and_enable_validators` further by having it instantiate a validator rather than taking one as as argument. In most cases we create a validator, immediately call `compose_...`, and then sometimes add some more rules. It is rare that we don't call `compose_...`.

Reply:
After a brief discussion we decided to leave it as is.